### PR TITLE
Arm/master/memsize passing v1

### PR DESCRIPTION
**Pull request by marex**
This patchset adds support for passing DRAM configuration and board compatible string from ATF to next stages.

### CODE REVIEW

**Comment by ldts** [comment]
thanks @marex. will try to do this before EOD (GMT) but all seems good to me.

File: plat/renesas/rcar/bl2_plat_setup.c [hunk 1]
```diff
@@ -85,6 +86,32 @@ CASSERT((PARAMS_BASE + sizeof(bl2_to_bl31_params_mem_t) + 0x100)
 
 static meminfo_t bl2_tzram_layout __aligned(CACHE_WRITEBACK_GRANULE);
 
+/* FDT with DRAM configuration */
+static uint64_t fdt_blob[4096 / sizeof(uint64_t)];
```
Comment (antonio-nino-diaz-arm) [inline plat/renesas/rcar/bl2_plat_setup.c]: 4096? Is this meant to be the page size, or is it just an arbitrary number?

File: plat/renesas/rcar/bl2_plat_setup.c [hunk 2]
```diff
@@ -314,6 +343,54 @@ static void bl2_advertise_dram_entries(uint64_t dram_config[8])
...
+		size = dram_config[2 * chan + 1];
+		if (!size)
+			continue;
+
+		/*
+		 * Channel 0 is mapped in 32bit space and the first
+		 * 128 MiB are reserved
+		 */
+		if (chan == 0) {
+			start = 0x48000000;
```
Comment (antonio-nino-diaz-arm) [inline plat/renesas/rcar/bl2_plat_setup.c]: It would be better to not use magic numbers.

File: plat/renesas/rcar/bl2_plat_setup.c (hunk 1 above)
Comment (marex) [reply plat/renesas/rcar/bl2_plat_setup.c]: Arbitrary number into which the DT fits. What better size than a page-aligned one.

File: plat/renesas/rcar/bl2_plat_setup.c (hunk 2 above)
Comment (marex) [reply plat/renesas/rcar/bl2_plat_setup.c]: The comment above explains those.

File: plat/renesas/rcar/bl2_plat_setup.c (hunk 1 above)
Comment (ldts) [reply plat/renesas/rcar/bl2_plat_setup.c]: why not then use PAGE_SIZE_4KB?

**Comment by marex** [comment]
On 12/05/2018 10:53 PM, ldts wrote:
> *@ldts* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In plat/renesas/rcar/bl2_plat_setup.c
> <https://github.com/ARM-software/arm-trusted-firmware/pull/1704#discussion_r239251899>:
> 
>> @@ -85,6 +86,32 @@ CASSERT((PARAMS_BASE + sizeof(bl2_to_bl31_params_mem_t) + 0x100)
>  
>  static meminfo_t bl2_tzram_layout __aligned(CACHE_WRITEBACK_GRANULE);
>  
> +/* FDT with DRAM configuration */
> +static uint64_t fdt_blob[4096 / sizeof(uint64_t)];
> 
> why not then use PAGE_SIZE_4KB?

Fixed

-- 
Best regards,
Marek Vasut

**Comment by marex** [comment]
On 12/06/2018 09:11 AM, ldts wrote:
> *@ldts* commented on this pull request.
> 
> it is not clear to me how the device tree makes it to bl33 in the
> ep_info.args. AFAICS you'd have to build with RCAR_BL33_ARG0 holding the
> value of the address? could you clarify please? I think we discussed
> this some weeks ago but I am not sure what you decided in the end.

The DT is being passed via arg3, except that code disappeared during the
rebase on the ATF master. Now my question is, where/how do I set the
arg3 in ATF/master ? The old function to do that is gone.

-- 
Best regards,
Marek Vasut

**Comment by ldts** [comment]
what function were you using before rebasing? I think you should be able to just pass the dt address with a change in bl2_plat_mem_params_desc.c (using arg3 instead of arg0)

**Comment by marex** [comment]
On 12/06/2018 09:58 AM, ldts wrote:
> what function were you using before rebasing? I think you should be able
> to just pass the dt address with a change in bl2_plat_mem_params_desc.c
> (using arg3 instead of arg0)

bl2_plat_set_bl33_ep_info() . I guess I can put it into the static
table, since the FDT buffer is also static chunk of memory.

-- 
Best regards,
Marek Vasut

**Comment by marex** [comment]
On 12/06/2018 10:26 AM, ldts wrote:
> maybe you can call get_bl_mem_params_node(BL33_IMAGE_ID) from
> bl2_plat_handle_pre_image_load and update the node there with the fdt
> address in arg3?

Maybe I can put it into the static bl2_mem_params_desc[] array , since
it's a static chunk of memory anyway ?

-- 
Best regards,
Marek Vasut

**Comment by ldts** [comment]
sounds good to me

**Comment by ldts** [comment]
how this going?

**Comment by ldts** [comment]
thanks Marex, looks good to me. @antonio-nino-diaz-arm ?

**Comment by antonio-nino-diaz-arm** [comment]
Looks good to me.

**Comment by antonio-nino-diaz-arm** [comment]
jenkins: test this please

**Comment by antonio-nino-diaz-arm** [comment]
The CI failures are due to merging some patches in TF-A-Tests, this PR is correct.

File: plat/renesas/rcar/bl2_plat_mem_params_desc.c [hunk 3]
```diff
@@ -19,6 +20,8 @@
 #define BL33_MODE MODE_EL2
 #endif
 
+extern uint64_t fdt_blob[PAGE_SIZE_4KB / sizeof(uint64_t)];
```
Comment (soby-mathew) [inline plat/renesas/rcar/bl2_plat_mem_params_desc.c]: Is the kernel DTB being created in secure memory ? How would BL33 access the secure memory ?

Also, would prefer to use arg1 for the kernel DTB to be inline with Dynamic config framework in TF-A. https://github.com/ARM-software/arm-trusted-firmware/blob/master/docs/firmware-design.rst#11dynamic-configuration-during-cold-boot

**Comment by marex** [comment]
On 12/10/2018 04:13 PM, Soby Mathew wrote:
> *@soby-mathew* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In plat/renesas/rcar/bl2_plat_mem_params_desc.c
> <https://github.com/ARM-software/arm-trusted-firmware/pull/1704#discussion_r240248769>:
> 
>> @@ -19,6 +20,8 @@
>  #define BL33_MODE MODE_EL2
>  #endif
>  
> +extern uint64_t fdt_blob[PAGE_SIZE_4KB / sizeof(uint64_t)];
> 
> Is the kernel DTB being created in secure memory ? How would BL33 access
> the secure memory ?

I can just read the memory address passed in x3 from U-Boot .

> Also, would prefer to use arg1 for the kernel DTB to be inline with
> Dynamic config framework in TF-A.
> https://github.com/ARM-software/arm-trusted-firmware/blob/master/docs/firmware-design.rst#11dynamic-configuration-during-cold-boot
Is this something we want even if the DT is just a small fragment and
not a complete DT ?

-- 
Best regards,
Marek Vasut

**Comment by soby-mathew** [comment]
> I can just read the memory address passed in x3 from U-Boot .

OK, I am not sure whether the fdt_blob[] is being copied to non-secure memory prior to handover to uboot. If x3 is pointing a memory within the secure address space (i.e within B2), and Uboot shouldn't be able to read it.

> Is this something we want even if the DT is just a small fragment and
not a complete DT ?

In this case, it can be either treated as config data to uboot (i.e NT_FW_CONFIG) or as kernel DT (HW_CONFIG). But I see you are already using x0 for other purpose, hence using x1 (HW_CONFIG) might be easier.  Although this is optional, it might be easier to do the changes now rather than later as this would become ABI between TF-A and UBoot for rcar.

**Comment by marex** [comment]
On 12/12/2018 10:49 AM, Soby Mathew wrote:
>     I can just read the memory address passed in x3 from U-Boot .
> 
> OK, I am not sure whether the fdt_blob[] is being copied to non-secure
> memory prior to handover to uboot. If x3 is pointing a memory within the
> secure address space (i.e within B2), and Uboot shouldn't be able to
> read it.
> 
>     Is this something we want even if the DT is just a small fragment and
>     not a complete DT ?
> 
> In this case, it can be either treated as config data to uboot (i.e
> NT_FW_CONFIG) or as kernel DT (HW_CONFIG). But I see you are already
> using x0 for other purpose, hence using x1 (HW_CONFIG) might be easier.

So like this ?
https://github.com/marex/arm-trusted-firmware-1/commit/06ce6912b314fc32b7971a424417429c633d3df6

> Although this is optional, it might be easier to do the changes now
> rather than later as this would become ABI between TF-A and UBoot for rcar.

It's fine by me either way, better be consistent.

-- 
Best regards,
Marek Vasut

**Comment by soby-mathew** [comment]
> So like this ?
> marex@06ce691

Yes, that should be OK. 

Also, could you clarify about the memory being pointed to ?  If it is memory within BL2, and it is accessible by BL33, it is a security violation and needs to be fixed (via suitable TrustZone settings).

**Comment by marex** [comment]
On 12/14/2018 01:19 PM, Soby Mathew wrote:
>     So like this ?
>     marex/arm-trusted-firmware-1@06ce691
>     <https://github.com/marex/arm-trusted-firmware-1/commit/06ce691>
> 
> Yes, that should be OK.

Fine, let me send a PR.

(I presume git request-pull is not supported by ATF or github, right?)

> Also, could you clarify about the memory being pointed to ? If it is
> memory within BL2, and it is accessible by BL33, it is a security
> violation and needs to be fixed (via suitable TrustZone settings).

What am I looking for exactly ? I'm still getting into this ATF stuff.

-- 
Best regards,
Marek Vasut

**Comment by soby-mathew** [comment]
A PR should be fine.

> What am I looking for exactly ? I'm still getting into this ATF stuff.

Basically all of the firmware is supposed to run in memory which is `Secure` access only. Software without `Secure` world privilege like U-Boot/Linux cannot access this memory.

The memory can be `secure` either via hardware configuration or runtime configurable by software. If configurable, this would be done with the aid of a TrustZone controller (TZC-400) or something similar. There are some security setting in `plat/renesas/rcar/bl2_secure_setting.c` but not sure whether this configures the security for system RAM as well.

**Comment by soby-mathew** [comment]
@AlexeiFedorov gave me some inputs, but we didn't figure out whether System RAM is Secure-access only or not.

**Comment by marex** [comment]
On 12/17/18 6:32 PM, Soby Mathew wrote:
> @AlexeiFedorov <https://github.com/AlexeiFedorov> gave me some inputs,
> but we didn't figure out whether System RAM is Secure-access only or not.

If I dump the xlat tables (v2) for the SRAM, it is marked as
MEM-RO-EXEC-S (and a chunk which is MEM-RW-XN-S), so I take it is secure
access only.

-- 
Best regards,
Marek Vasut

**Pull request merged**